API for storing metadata

kazcw opened the pull request:
### Pull Request Description

API for storing metadata.

See: https://www.pivotaltracker.com/story/show/181149277

### Important Notes

**New APIs**:
- Storing metadata is implemented with `profiler::MetadataLogger`.
- A full metadata storage/retrieval example is in [the top-level doctests](https://github.com/enso-org/enso/blob/wip/kw/profiling-metadata-api/lib/rust/profiler/data/src/lib.rs) for profiler::data, a crate which implements an API for profiling data consumers (it abstracts away the low-level details of the event log, and checks its invariants in the process) [after review of this new API here I'll open a PR to add it to the design doc].

**Implementation**:
- `profiler::Event` is parameterized by a metadata type, so that different types of metadata can be dependency-injected into it.
- A data consumer defines its metadata type as an enum of all the kinds of metadata it is interested in.
- Producing the metadata enum is accomplished without defining its type (which would require dependencies from around the app): A `MetadataLogger` internally use a serialization helper `Variant` to serialize its variant of the metadata enum without knowledge of the other possible variants.

**Performance impact**: still in the low ns/measurement range, comparable to pushing to a vec.

*Note*: `LocalVecBuilder` is currently present under the name `Log`, which is accurate but probably too overloaded. I'd like to find the right name for it, document it with examples, and move it to its own crate under data-structures, but I don't want doing that to hold up this PR.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

[ci no changelog needed]

kazcw committed:
API for storing metadata

See: https://www.pivotaltracker.com/story/show/181149277

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-metadata-api

MichaelMauderer commented on lib/rust/profiler/src/lib.rs:
```diff
@@ -431,21 +542,165 @@ pub enum Event {
+    /// Create a MetadataLogger for logging a particular type.
+    ///
+    /// The name given here must match the name used for deserialization.
+    pub fn register(name: &'static str) -> Self {
```
I wonder whether this could not just be called `new`. From an API user perspective, I think it would be clearer to see that this is the default way to instantiate a `MetadataLogger`.

kazcw replied:
I wanted to convey that it's a side-effectful operation, but that's probably obvious enough. I'll change it.

MichaelMauderer commented on lib/rust/profiler/data/src/lib.rs:
```diff
@@ -0,0 +1,686 @@
+//!     demo();
+//!
+//!     #[derive(serde::Deserialize, PartialEq, Eq, Debug)]
+//!     enum MyMetadata {
```
I think this enum might require a bit more explanation. How is this matched to the serialized data? I assume internally this is matched on the JSON representation, but this is not obvious in this example.

kazcw replied:
When I was looking for a good place to document this in the API, I realized `take_log<Metadata>`/`interpret` is a poor entry point to this API. I've changed it so `take_log` now returns the log as a JSON `String`, and `Measurement::parse` interprets a log. Then there's a place to document how the metadata enum is used: `Measurement<Metadata>`.

Also, I hadn't implemented error-recovery for metadata changes, so I added that and a `skip_failed_metadata` test case.

MichaelMauderer commented on lib/rust/profiler/src/lib.rs:
```diff
@@ -746,6 +1013,51 @@ mod tests {
     }
 }
 
+#[cfg(test)]
```
It would be nice to have an example/test that shows the JSON output and the result of parsing it. 
Something like serializing some data, asserting it against the string, and deserializing the sample  string and asserting the resulting data.

kazcw replied:
Added a `format_stability` test. In order to avoid the test depending on the exact definitions of event types, it is a deserialize-serialize test rather than a serialize-deserialize test.

MichaelMauderer commented on lib/rust/profiler/src/log.rs:
```diff
@@ -0,0 +1,119 @@
+    }
+}
+
+// Note [Log Safety]
```
I wonder whether this comment should not be in a more prominent place towards the beginning of the code using the unsafe functionality, or even as part of the struct docs.

kazcw committed:
parse() API, error recovery, format test, misc

Misc:
- MetadataLogger::new
- document usage of metadata enum
- move docs for log safety

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-metadata-api

kazcw commented:
@MichaelMauderer While commenting on #3297, I realized data consumers should have an easier way to distinguish async from non-async Measurements. I'll revise that now...

kazcw committed:
Distinguish sync/async lifetimes

kazcw commented:
A `Measurement` now has a `Lifetime` enum that can be used to handle sync/async measurements differently; async measurements have a `create_to_finish` convenience method that we can use when graphing to show a profiler's total extent.

kazcw committed:
create_to_finish: fix edge cases + add tests

wdanilo commented on lib/rust/profiler/src/lib.rs:
```diff
@@ -246,16 +271,80 @@ pub mod js {
+}
+
+
+// === mod id ===
+
+/// Facilities for interpreting an event ID.
```
Could you extend this docs please? It's worth describing what kin of interpretation does it handle and provide to the reader a little more context.

wdanilo commented on lib/rust/profiler/src/lib.rs:
```diff
@@ -431,21 +531,165 @@ pub enum Event {
+    Metadata(Metadata<M>),
+}
+
+impl<M, L> Event<M, L> {
+    /// Produce a new event that may have a different metadata type, with metadata values
+    /// converted by the given function.
+    fn map_metadata<F, M1>(self, mut f: F) -> Event<M1, L>
+    where F: FnMut(M) -> M1 {
+        match self {
+            Event::Start(start) => Event::Start(start),
+            Event::StartPaused(start) => Event::StartPaused(start),
+            Event::End { id, timestamp } => Event::End { id, timestamp },
+            Event::Pause { id, timestamp } => Event::Pause { id, timestamp },
+            Event::Resume { id, timestamp } => Event::Resume { id, timestamp },
+            Event::Metadata(Metadata { timestamp, data }) =>
+                Event::Metadata(Metadata { timestamp, data: f(data) }),
+        }
+    }
```
Sometimes it's worth making type ariables longer. I believe this is the case. `M` and `L` and not very descriptive and its hard to understand what they mean. Alternatively, this can be provided in short docs.

wdanilo commented on lib/rust/profiler/src/lib.rs:
```diff
@@ -431,21 +531,165 @@ pub enum Event {
+
+// === OpaqueMetadata ===
+
+/// Black-box metadata object, for use when metadata doesn't need to be created/interpreted.
+#[derive(Debug, Copy, Clone, serde::Serialize, serde::Deserialize)]
+pub enum OpaqueMetadata {
```
Could you extend these docs slightly, please? What I'm missing is info what use cases does it cover – when the metadata doesn't need to be interpreted?

kazcw replied:
Used only for testing now, so I made it a private item in `tests`.

wdanilo reviewed (approved):
I really like the code. It's unusual that for such a big PR I have only a few comments :P Anyway, I have three additional questions/requests:
1. Could you confirm that if we are not using metadata this does not affect the profiler performance?
2. Some of the files are getting really big. Do you think we could refactor them down to sub-modules, or you feel that it will make them less readable?
3. I feel that docs need some more explanation here and there – could you go over each file and read it from top to bottom, trying to read it from a perspective someone who does not know this design at all? Please be sure that after reading the docs from top to bottom the reader has full and clear picture of the design. I'm asking about it normally in the final stage of the implementation of the library, and it seems that we are getting there :)

Anyway, a green light from me. I'd be thankful if you'd apply the points above and feel free to merge it :)

wdanilo commented:
Btw, pls add me to codeowners of libs - this way I'd be notified about this PR earlier and I'd review it earlier.

kazcw committed:
CODEOWNERS: add @wdanilo for /lib/rust/

kazcw commented:
@wdanilo Re: performance

I don't expect any performance change. Benchmarks are the same within variance.

#### BEFORE

test bench::log_measurements_1000   ... bench:       3,892 ns/iter (+/- 581)
test bench::log_measurements_10_000 ... bench:      36,740 ns/iter (+/- 290)

#### AFTER

test bench::log_measurements_1000   ... bench:       3,679 ns/iter (+/- 79)
test bench::log_measurements_10_000 ... bench:      37,063 ns/iter (+/- 34,044)
test bench::push_vec_1000           ... bench:       1,997 ns/iter (+/- 74)
test bench::push_vec_10_000         ... bench:      20,100 ns/iter (+/- 651)

kazcw committed:
Reorganize, document

kazcw commented:
### Modules

#### `profiler`

Moved implementation detalis into `profiler::internal`; now the top-level contents of `profiler` contain only the API intended for general use (and described in the design doc). `profiler::internal` is used to implement the macros, and used by `profiler::data` to decode raw event logs.

#### `profiler::data`

Moved parsing implementation into `profiler::data::parse`. The top-level module now only defines the interface.

kazcw committed:
Rename Metadata to Timestamped; rename type vars.

kazcw committed:
Docs & organization

kazcw committed:
Simplify some MeasurementBuilder logic

wdanilo commented:
@kazcw please never do this:
<img width="799" alt="image" src="https://user-images.githubusercontent.com/1623053/155800704-c6a3f89c-a0bd-42f7-b4ca-c9fc99a5b5a8.png">

Now, when browsing the code, I cant see old discussions and my old review comments.

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-metadata-api

kazcw commented:
This is awaiting approval by a codeowner of the base branch. @MichaelMauderer?

MichaelMauderer commented:
> This is awaiting approval by a codeowner of the base branch. @MichaelMauderer?

Approved.
